File: programming-paradigm/edp/js/event_driven.js
/*
一个简单的事件驱动系统，包含事件的注册、触发和移除功能。不同语言的实现方式有所不同，但核心逻辑是一致的：
事件管理器：负责管理事件及其对应的回调函数列表。
注册事件：将回调函数添加到对应事件的列表中。
触发事件：遍历对应事件的回调函数列表并依次调用。
移除事件：从对应事件的回调函数列表中移除指定的回调函数。
*/
// 自定义事件管理器类
class EventManager {
  constructor() {
    // 用于存储事件及其对应的回调函数列表
    this.events = {};
  }

  // 注册事件监听器的方法
  on(eventName, callback) {
    if (!this.events[eventName]) {
      // 如果该事件还没有对应的回调函数列表，就初始化一个空数组
      this.events[eventName] = [];
    }
    // 将回调函数添加到对应事件的回调列表中
    this.events[eventName].push(callback);
  }

  // 触发事件的方法
  emit(eventName, ...args) {
    if (this.events[eventName]) {
      // 遍历该事件对应的所有回调函数
      this.events[eventName].forEach(callback => {
        // 依次调用每个回调函数，并传入参数
        callback(...args);
      });
    }
  }

  // 移除事件监听器的方法
  off(eventName, callback) {
    if (this.events[eventName]) {
      // 过滤掉指定的回调函数，保留其他回调函数
      this.events[eventName] = this.events[eventName].filter(cb => cb !== callback);
    }
  }
}

// 使用示例
// 创建事件管理器实例
const eventManager = new EventManager();

// 定义事件处理函数
function handleMessage(message) {
  console.log(`收到消息: ${message}`);
}

(function test(params) {
  // 第一次测试：注册并触发事件
  console.log("第一次测试：");
  // 注册事件监听器
  eventManager.on('message', handleMessage);
  // 触发事件
  eventManager.emit('message', '第一次发送的消息');

  // 第二次测试：再次触发事件
  console.log("\n第二次测试：");
  eventManager.emit('message', '第二次发送的消息');

  // 第三次测试：移除监听器后重新注册并触发事件
  console.log("\n第三次测试：");
  // 移除事件监听器
  eventManager.off('message', handleMessage);
  // 重新注册事件监听器
  eventManager.on('message', handleMessage);
  // 触发事件
  eventManager.emit('message', '第三次发送的消息');
})()

/*
jarry@Mac js % node event_driven.js
第一次测试：
收到消息: 第一次发送的消息

第二次测试：
收到消息: 第二次发送的消息

第三次测试：
收到消息: 第三次发送的消息
*/

File: programming-paradigm/edp/go/event_driven.go
package main

import "fmt"

/*
一个简单的事件驱动系统，包含事件的注册、触发和移除功能。不同语言的实现方式有所不同，但核心逻辑是一致的：
事件管理器：负责管理事件及其对应的回调函数列表。
注册事件：将回调函数添加到对应事件的列表中。
触发事件：遍历对应事件的回调函数列表并依次调用。
移除事件：从对应事件的回调函数列表中移除指定的回调函数。
*/
// 事件管理器结构体
type EventManager struct {
  events map[int][]func(string)
}

// 初始化事件管理器
func NewEventManager() *EventManager {
  return &EventManager{
    events: make(map[int][]func(string)),
  }
}

// 注册事件监听器
func (em *EventManager) On(eventName int, callback func(string)) {
  if _, exists := em.events[eventName]; !exists {
    em.events[eventName] = make([]func(string), 0)
  }
  em.events[eventName] = append(em.events[eventName], callback)
}

// 触发事件
func (em *EventManager) Emit(eventName int, message string) {
  if callbacks, exists := em.events[eventName]; exists {
    for _, callback := range callbacks {
      callback(message)
    }
  }
}

// 移除事件监听器
func (em *EventManager) Off(eventName int, callback func(string)) {
  if callbacks, exists := em.events[eventName]; exists {
    for i, cb := range callbacks {
      if fmt.Sprintf("%p", cb) == fmt.Sprintf("%p", callback) {
        em.events[eventName] = append(callbacks[:i], callbacks[i+1:]...)
        break
      }
    }
  }
}

// 事件处理函数
func handleMessage(message string) {
  fmt.Printf("收到消息: %s\n", message)
}

func main() {
  eventManager := NewEventManager()

  // 第一次测试：注册并触发事件
  fmt.Println("第一次测试：")
  // 注册事件监听器
  eventManager.On(0, handleMessage)
  // 触发事件
  eventManager.Emit(0, "第一次发送的消息")

  // 第二次测试：再次触发事件
  fmt.Println("\n第二次测试：")
  eventManager.Emit(0, "第二次发送的消息")

  // 第三次测试：移除监听器后重新注册并触发事件
  fmt.Println("\n第三次测试：")
  // 移除事件监听器
  eventManager.Off(0, handleMessage)
  // 重新注册事件监听器
  eventManager.On(0, handleMessage)
  // 触发事件
  eventManager.Emit(0, "第三次发送的消息")
}

/*
jarry@Mac go % go run event_driven.go
第一次测试：
收到消息: 第一次发送的消息

第二次测试：
收到消息: 第二次发送的消息

第三次测试：
收到消息: 第三次发送的消息
*/
